Split shop buttons

nickshanks opened the pull request:
# Feature

## Summary

* Split the shop panel buttons into two halves, shop inventory buttons on the left, fleet management on the right.
* Add a Park button.
* Add a selected ship count (not in screenshot; selection >= 2).
* Move the definitons into the interfaces file.

Moving the Find button, player credits, and ship selection text into the interfaces file are left as excercises for the reader.

There is an existing bug whereby clicking and dragging in the buttons area but not on any button, drags the active pane. I was not able to find and fix this.

## Screenshots
<img width="1312" alt="Screenshot 2024-06-10 at 15 11 28" src="https://github.com/endless-sky/endless-sky/assets/206120/c76eeb7f-271c-420d-88b4-f879e0bbc4e7">
<img width="1312" alt="Screenshot 2024-06-10 at 15 16 42" src="https://github.com/endless-sky/endless-sky/assets/206120/bded74ec-79c2-4d1c-9fed-7d82e8736efd">

## Testing Done
I think I checked all possible combinations of button label and behaviour, but I would appreciate some third party testing as well.

## Save File
Any save file will do.

## Wiki Update
I don't really know what's in the wiki. Maybe a screenshot or text update is needed on some page?

## Performance Impact
N/A

TomGoodIdea commented:
You're adding a single button that essentially duplicates some functionality that is already accessible from the Player Info panel, which additionally results in shop buttons taking up twice as much space as before. I'm not a fan of this.

nickshanks commented:
This is a stepping stone towards splitting the cargo, install, storage features, which presently are all overloaded onto the Buy button. You currently cannot move an item to cargo without Command-clicking the currently selected ship to deselect it. And there is a nuanced difference between the behaviour of "i" vs "b" which is not available via the mouse.

It is also a clearer logical separation of buying/selling the fungible items on the left from non-fungible ships in your fleet.

TomGoodIdea commented on data/_ui/interfaces.txt:
```diff
@@ -1292,6 +1293,81 @@ interface "bank"
+		dimensions 70 30
+		size 18
+
+	visible if "is outfitters"
```
Shouldn't be singular?
```suggestion
	visible if "is outfitter"
```

nickshanks replied:
I would agree but the precedent has already been set and I didn't want to use two different values in two places in interfaces.txt as that makes maintenance and plugins harder.

TomGoodIdea replied:
Precedent? Where?

TomGoodIdea commented on data/_ui/interfaces.txt:
```diff
@@ -1292,6 +1293,81 @@ interface "bank"
+		dimensions 70 30
+		size 18
+
+	visible if "!is outfitters"
```
```suggestion
	visible if "!is outfitter"
```

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -847,72 +864,75 @@ void ShopPanel::DrawDetailsSidebar()
+	const bool isOwned = IsAlreadyOwned();
+	info.SetCondition(isOwned ? (playerShip ? "use install" : "use cargo") : "use buy");
+	if(isOutfitter)
+		info.SetCondition("is outfitters");
```
```suggestion
		info.SetCondition("is outfitter");
```

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -388,6 +390,21 @@ bool ShopPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, boo
+		const Ship *flagship = player.Flagship();
+		for(Ship *ship : playerShips)
+		{
+			if(!ship->IsDisabled() && ship != flagship)
```
Can you have disabled ships in the shop?

nickshanks replied:
No, they don't show up when in system but disabled. I think this was there originally just for defensive purposes (there's a lot of "this shouldn't happen but just in case…" code).
It was there originally so I left it.

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -847,72 +864,75 @@ void ShopPanel::DrawDetailsSidebar()
+		font.Draw(to_string(playerShips.size()) + " ships selected", textPoint, dim);
+	}
+
+	const Interface *shopUi = GameData::Interfaces().Get("shop");
```
shopUi isn't a `UI`. It's an `Interface`, so please rename it.

nickshanks replied:
99% of other places where Interface::Get() is called, the result is assigned to a variable called fooUi. Why pick on this one to be different?

nickshanks replied:
I consider it a design error to have a class called Interface and a second class called UI. I propose renaming UI to PanelStack.

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -1458,28 +1528,28 @@ vector<ShopPanel::Zone>::const_iterator ShopPanel::Selected() const
 {
-	if(x > Screen::Right() - SIDEBAR_WIDTH - 342 && x < Screen::Right() - SIDEBAR_WIDTH - 316 &&
+	const Point click(x, y);
+	const Interface *shopUi = GameData::Interfaces().Get("shop");
```
same as above

nickshanks replied:
See above

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -1458,28 +1528,28 @@ vector<ShopPanel::Zone>::const_iterator ShopPanel::Selected() const
+	const Rectangle buyButton = shopUi->GetBox("buy button");
+	const Rectangle sellButton = shopUi->GetBox(isOutfitter ? "sell button" : "sell ship button");
+
+	if(leaveButton.Contains(click))
+		return 'l';
+	if(parkButton.Contains(click))
+		return 'k';
+	if(buyButton.Contains(click))
+		return IsAlreadyOwned() ? 'i' : 'b';
+	if(sellButton.Contains(click))
+		return 's';
```
Why not just use buttons if you're already moving this to an interface?

nickshanks replied:
I haven't figured out how to make their click zones work automatically. If you look in interfaces.txt they ARE buttons, but i could not click them.

Please help if you know what I missed or did wrong.

TomGoodIdea replied:
Ah, I guess it needs to be done your way if we want the keypress to be sent even when the button is inactive. But if you have both a button and a rectangle zone, doesn't it trigger twice on a single click?

nickshanks replied:
Well if it weren't for the problem that the button definitions don't seem to work, I would have thought the same.
The whole DoButtons() function should be unnecessary but I couldn't get buttons to work here the way they do in other panel Interfaces that use "button".

TomGoodIdea commented:
https://github.com/endless-sky/endless-sky/pull/10187#discussion_r1635899609
@nickshanks can you please explain where it's already used?

TomGoodIdea commented:
There is a difference between a **single outfitter** on a planet, and a map that collects info from **multiple outfitters**.

nickshanks commented:
> There is a difference between a **single outfitter** on a planet, and a map that collects info from **multiple outfitters**.

True, and I said they were for different things, but I still think it's better not to introduce a new key that only differed by one letter for a very similar concept.

TomGoodIdea commented:
It doesn't make sense to name it "outfitters" if everywhere else the game calls a single outfitter an "outfitter".

nickshanks commented:
> It doesn't make sense to name it "outfitters" if everywhere else the game calls a single outfitter an "outfitter".

I've already had this struggle internally before committing any code. If it helps, perhaps think of it as being in the genitive.

I'd like the other devs to weigh in. I see it as keeping the API consistent, over being grammatically precise.

TomGoodIdea commented:
If you pass `this` instead of a nullptr to the interface draw function, it should register the buttons properly.

nickshanks commented:
> If you pass `this` instead of a nullptr to the interface draw function, it should register the buttons properly.

Ahhhhhh! So that's how it works.
I updated the PR to do this, and remove the click handlers for buttons defined in the Interface. I also moved the Find button into the interface. I tried to eliminate CheckButton() but that caused tooltips for hidden table rows/player ships to show up when hovering over the buttons area (including two tooltips simultaneously if hovering over the credits value!)

Some improvements I would like to make (in other PRs):

* Use the `box` definitions to define a button's position and extent.
* Specify an icon (incl. hover state) for a button.

TomGoodIdea commented:
Waiting for a Dev to resolve the grammar problem

tibetiroka commented:
I'm not a huge fan of calling it `is outfitters`, especially that the variable in the code is called `isOutfitter`. We are on a different panel, so we don't need to reuse the condition names, especially when they don't mean the same thing.

nickshanks commented:
I'll take a look at the weekend.

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -836,6 +854,7 @@ void ShopPanel::DrawShipsSidebar()
 		font.Draw({space, {SIDEBAR_WIDTH - 20, Alignment::RIGHT}}, point, bright);
 		point.Y() += 20.;
 	}
+
```
```suggestion
```

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -889,79 +908,80 @@ void ShopPanel::DrawDetailsSidebar()
+	}
+
+	const Interface *shopUi = GameData::Interfaces().Get("shop");
+	const Rectangle leaveButton = shopUi->GetBox("leave button");
+	const Rectangle parkButton = shopUi->GetBox("park button");
+	const Rectangle buyButton = shopUi->GetBox("buy button");
+	const Rectangle sellButton = shopUi->GetBox(isOutfitter ? "sell button" : "sell ship button");
```
If we had a method to access a button element and another to get its bounds, this button-box duplication wouldn't be needed, but I guess it's fine for now.

TomGoodIdea reviewed (approved):
seems to work

tibetiroka commented on source/ShopPanel.h:
```diff
@@ -210,10 +210,12 @@ class ShopPanel : public Panel {
-	// letter of the button (or ' ' if it's not on a button).
 	char CheckButton(int x, int y);
 
+	bool EscortSelected();
+	bool CanPark();
+	bool CanUnpark();
```
This could use some comments.

```suggestion
	// Check if at least one of the selected ships is an escort.
	bool EscortSelected();
	// Check whether there is a selected ship that can be parked/unparked.
	bool CanPark();
	bool CanUnpark();
```

tibetiroka commented on data/_ui/interfaces.txt:
```diff
@@ -1298,6 +1299,88 @@ interface "bank"
+	active if "can sell"
+	button s "_Sell"
+		center -215 -25
+		dimensions 50 30
```
```suggestion
		dimensions 70 30
```

nickshanks replied:
I'm happy to make the Sell buttons match the other button size, but the existing width is 50px which I presume is to help avoid accidental clicks, so I didn't want to change that as part of this. Are you sure you want me to change this?

tibetiroka replied:
Yes. It looks more awkward now that it has buttons on both sides, and we have a popup confirmation dialog that should help avoid accidentally selling ships.

tibetiroka commented on data/_ui/interfaces.txt:
```diff
@@ -1298,6 +1299,88 @@ interface "bank"
+		dimensions 70 30
+	box "sell ship button"
+		center -215 -25
+		dimensions 50 30
```
```suggestion
		dimensions 70 30
```

nickshanks replied:
you'd also need to move the centre of the box by 10px to compensate for the 20px wider button

nickshanks replied:
I have to go to work now. I'll try and position the wider buttons tonight.

tibetiroka replied:
It looks like they are already slightly offset compared to the buttons on the right side, so maybe they were positioned for their full size?

TomGoodIdea commented:
there are merge conflicts
